### PR TITLE
Implements a Node Parser using embeddings for Semantic Splitting

### PR DESCRIPTION
**Pull request by DustinEwan**
# Description

Provided an implementation of BaseEmbedding, allows Documents to be split into nodes based on semantic similarity. First, the text is split into sentences using the provided sentence splitter.  Then, sentences are recursively grouped together to form a list of all possible continguous sentence permutations in the document.

Embeddings are then generated for each individual sentence and each group of sentences, then the cosine distance is calculated to determine the similarity between each group of sentence and the following one. Percentiles are then calculated for each similarity to normalize the distribution and identify outliers.

Finally, each similarity percentile is evaluated against a user-provided threshold to determine breakpoints for semantic groups of sentences.  Providing a higher breakpoint percentile will result in larger nodes, and visa versa.  The default value of 95th percentile seems to be a good default in most scenarios.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/node_parser/text/semantic_splitter.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+class SentenceCombination(TypedDict):
+    sentence: str
+    index: int
+    combined_sentence: str
+    combined_sentence_embedding: List[float]
+    distance_to_next: float
+
+
+class SemanticSplitterNodeParser(NodeParser):
```
Comment (logan-markewich) [inline llama_index/node_parser/text/semantic_splitter.py:26]: I wonder if we can add a unit test or two for this. Maybe by mocking the embedding model 🤔

Comment (logan-markewich) [reply llama_index/node_parser/text/semantic_splitter.py:26]: Or embeddings aside, we could refactor the `build_semantic_nodes_from_documents()` into a few different functions to make the overall logic more testable

Comment (DustinEwan) [reply llama_index/node_parser/text/semantic_splitter.py:26]: Yeah, I was wondering the same.  I actually just pinged you on Discord mentioning this :)

File: llama_index/node_parser/text/semantic_splitter.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                combined_sentence += sentences[i]["sentence"]
+
+                for j in range(i + 1, i + 1 + self.window_size):
+                    if j < len(sentences):
+                        combined_sentence += " " + sentences[j]["sentence"]
+
+                sentences[i]["combined_sentence"] = combined_sentence
+                sentences[i][
+                    "combined_sentence_embedding"
+                ] = self.embedding.get_text_embedding(combined_sentence)
```
[the remark this answers is not in the document]
Comment (DustinEwan) [reply llama_index/node_parser/text/semantic_splitter.py:141]: testing out this change now

File: llama_index/node_parser/text/semantic_splitter.py (hunk 1 above)
Comment (DustinEwan) [reply llama_index/node_parser/text/semantic_splitter.py:26]: ah, yeah, that's fair on splitting it up :+1:

File: llama_index/node_parser/text/semantic_splitter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DustinEwan) [reply llama_index/node_parser/text/semantic_splitter.py:141]: oh... yeah... that's faster, lol.

**Comment by logan-markewich** [comment]
Made a few minor changes + added docs. Great work, and thanks for the contribution!!

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**